ios/RCTContacts/RCTContacts.m

```diff
@@ -61,41 +61,132 @@ - (NSDictionary *)constantsToExport
+
+    NSArray *nameKeys = @[
+        CNContactFamilyNameKey,
+        CNContactGivenNameKey,
+        CNContactMiddleNameKey,
+        CNContactNamePrefixKey,
+        CNContactNameSuffixKey,
+        CNContactPreviousFamilyNameKey,
+        CNContactNicknameKey,
+        CNContactPhoneticGivenNameKey,
+        CNContactPhoneticMiddleNameKey,
+        CNContactPhoneticFamilyNameKey
+        ];
+
+    NSMutableArray *keys = [[NSMutableArray alloc]init];
+    [keys addObjectsFromArray:nameKeys];
+    [keys addObjectsFromArray:@[
+        CNContactEmailAddressesKey,
+        CNContactPhoneNumbersKey,
+        CNContactPostalAddressesKey,
+        CNContactOrganizationNameKey,
+        CNContactJobTitleKey,
+        CNContactImageDataAvailableKey,
+        CNContactNoteKey,
+        CNContactUrlAddressesKey,
+        CNContactBirthdayKey
+        ]];
+
+    if (withThumbnails) {
+        [keys addObject:CNContactThumbnailImageDataKey];
+    }
+
+    CNContactFetchRequest *fetchRequest = [[CNContactFetchRequest alloc] initWithKeysToFetch:keys];
+    [fetchRequest setUnifyResults:YES];
+
+    NSError *regExError = nil;
+    NSRegularExpression *wordRegex = [NSRegularExpression regularExpressionWithPattern:@"\\w+"
+                                                                               options:0
+                                                                                 error:&regExError];
+    NSRegularExpression *nonNumberRegex = [NSRegularExpression regularExpressionWithPattern:@"\\D"
+                                                                                    options:0
+                                                                                      error:&regExError];
+
+    NSMutableArray<NSString*> *searchTerms = [[NSMutableArray alloc]init];
+    for (NSTextCheckingResult *match in [wordRegex matchesInString:searchString options:0 range:NSMakeRange(0, [searchString length])]) {
+        [searchTerms addObject:[searchString substringWithRange:[match range]]];
+    };
+    
+    [store enumerateContactsWithFetchRequest:fetchRequest
+                                       error:&contactError
+                                  usingBlock:^(CNContact * __nonnull contact, BOOL * __nonnull stop) {
+        // names are matched from the beginning only. Phone numbers are matched anywhere
+        NSMutableSet<NSString*> *names = [[NSMutableSet alloc] init];
+        NSMutableSet<NSString*> *contactPhoneNumbers = [[NSMutableSet alloc] init];
+
+        for (NSString *nameKey in nameKeys) {
+            NSString *name = [contact valueForKey:nameKey];
+            if (name) {
+                for (NSTextCheckingResult *matchObject in [wordRegex matchesInString:name options:0 range:NSMakeRange(0, name.length)]) {
+                    [names addObject:[name substringWithRange:[matchObject range]]];
+                }
+                [names addObject:name];
+            }
+        }
+
+        BOOL allTermsMatch = YES;
```

Comment:
This weird match vs allTermsMatch construct seems to be an inner loop break issue. Am I right?

I know everyone says dont use goto, but if so, that is a case where it may work.

Reply:
I think the `nextTerm` and `nextContact` labels are pretty easy to follow. I guess I can mark "use goto" off my weird programming checklist